### PR TITLE
Remove generic lifetime parameter of trait `Pattern`

### PR DESCRIPTION
**Pull request by a1phyr**
Use a GAT for `Searcher` associated type because this trait is always implemented for every lifetime anyway.

cc #27721

### CODE REVIEW

**Comment by compiler-errors** [comment]
Is this possible to change these method's signatures given the stability of libstd? This is an early-bound lifetime, so there may be (idk why they would, lol) code in the wild where they're turbofishing the lifetime.

**Comment by Amanieu** [comment]
@craterbot check

**Comment by compiler-errors** [comment]
@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-127481/retry-regressed-list.txt

**Comment by dtolnay** [comment]
Of the 54 regressions, 51 are in crates that use `#![feature(pattern)]` and can be discounted.

This one is worth reproducing locally to figure out what happened:

- https://crater-reports.s3.amazonaws.com/pr-127481-1/try%2381417ccfba1b50f064bed768ddab670f5bbfdcfe/gh/lcdr.lu_packets/log.txt (https://github.com/lcdr/lu_packets)

These 2 are legit regressions in stable Rust, as far as I can tell. This is a consequence of the change from early bound to late bound lifetime in the signatures.

- https://crater-reports.s3.amazonaws.com/pr-127481-1/try%2381417ccfba1b50f064bed768ddab670f5bbfdcfe/gh/Pulau-Komodo.chatbot/log.txt
- https://crater-reports.s3.amazonaws.com/pr-127481-1/try%2381417ccfba1b50f064bed768ddab670f5bbfdcfe/gh/rakurtz.aoc_2023/log.txt

```console
error[E0308]: mismatched types
 --> src/discord_client.rs:9:37
  |
9 |     let new_text = [str::strip_prefix, str::strip_suffix]
  |                                        ^^^^^^^^^^^^^^^^^ one type is more general than the other
  |
  = note: expected fn pointer `for<'a> fn(&'a _, &std::string::String) -> std::option::Option<&'a _>`
             found fn pointer `fn(&_, &std::string::String) -> std::option::Option<&_>`
```

**Comment by compiler-errors** [comment]
lu_packets is perma-broken and shows up on every crater report. It has nothing to do w/ this PR.

File: library/core/src/str/mod.rs [hunk 1]
```diff
@@ -1297,7 +1297,7 @@ impl str {
     #[inline]
     pub fn rfind<'a, P>(&'a self, pat: P) -> Option<usize>
     where
-        P: Pattern<'a, Searcher: ReverseSearcher<'a>>,
+        P: Pattern<Searcher<'a>: ReverseSearcher<'a>>,
```
Comment (Amanieu) [inline library/core/src/str/mod.rs:1300]: Would something like this help?
```suggestion
        P: Pattern,
        for<'a> P::Searcher<'a>: ReverseSearcher<'a>,
```

Comment (a1phyr) [reply library/core/src/str/mod.rs]: It should work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=af8eb63e3403dbfaaea02f2246a68bb6

**Comment by Amanieu** [comment]
This seems to have fixed all the stable regressions. The only remaining crates are the ones that use the unstable `Pattern` ABI and therefore aren't subject to our stability guarantees.

I noticed that several crates use build.rs to detect whether they are on a nightly compiler and automatically enable certain nightly features if that is the case.

### PLEASE STOP DOING THAT!

Nightly features should be opt-in via a Cargo feature, otherwise you end up breaking all users of your crate on nightly when we inevitably make breaking changes on unstable APIs. They are unstable for a reason!

@bors r+

**Pull request merged**